gen/field.go

```diff
@@ -94,23 +107,66 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) (string, error) {
+	ts := []string{}
+
+	jsonTag := compileJSONTag(f, f.Name)
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.Annotations[goTagKey]; goAnnotation != "" {
+		tags, err := structtag.Parse(goAnnotation)
+		if err != nil {
+			return "", errors.New(fmt.Sprintf("failed to parse tag %q: %v", goAnnotation, err))
+		}
+		for _, t := range tags.Tags() {
+			if t.Key == jsonTagKey {
+				jsonTag = compileJSONTag(f, t.Name, t.Options...)
+				continue
+			}
+			ts = append(ts, t.String())
+		}
+	}
+
+	ts = append(ts, jsonTag)
+
+	sort.Slice(ts, func(i, j int) bool {
+		return ts[i] < ts[j]
+	})
+
+	// TODO (abg): Take js.name annotation into account
+	return fmt.Sprintf("`%s`", strings.Join(ts, " ")), nil
```

Comment:
Optional: Some of this logic can further be simplified by using `*structtag.Tags` directly.

```go
tags, err := structtag.Parse("")  // no tags
if err != nil { .. }

if err := tags.Set(compileJSONTag(..)); err != nil { .. }
```

You won't need to have special handling for overwriting the json tag because
`tags.Set` will take care of stomping on it if it already exists.

And on this line, you can simply do,

```
fmt.Sprintf("`%s`", tags.String())
```